experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,699 @@
+    for a in f
+      parent(a) == base_ring(OO(X)) || error("element does not belong to the correct ring")
+    end
+    if length(name) > 0 
+      U = new{BRT, BRET, RT, RET, MST}(X, f)
+      set_name!(U, name)
+      return U
+    end
+    return new{BRT, BRET, RT, RET, MST}(X, f)
+  end
+end
+
+@Markdown.doc """
+    ambient(U::SpecOpen)
+
+Return the ambient scheme ``X`` of a Zariski open subset ``U ⊂ X``.
+"""
+ambient(U::SpecOpen) = U.X
+
+@Markdown.doc """
+    npatches(U::SpecOpen)
+
+Return the number of generators stored for describing the complement of ``U``.
+"""
+npatches(U::SpecOpen) = length(U.gens)
+
+@Markdown.doc """
+    gens(U::SpecOpen) = U.gens
+
+Return the generators ``[f₁,…,fᵣ]`` stored for the description 
+of the complement of ``U``.
+"""
+gens(U::SpecOpen) = U.gens
+
+@Markdown.doc """
+    getindex(U::SpecOpen, i::Int) = patches(U)[i]
+
+Return the hypersurface complement of ``fᵢ`` in the 
+ambient scheme ``X`` of ``U`` where ``f₁,…,fᵣ`` are 
+the generators stored for the description of the complement 
+of ``U``.
+"""
+getindex(U::SpecOpen, i::Int) = affine_patches(U)[i]
+
+function Base.show(io::IO, U::SpecOpen)
+  if isdefined(U, :name) 
+    print(io, name(U))
+    return
+  end
+  print(io, "complement of zero locus of $(gens(U)) in $(ambient(U))")
+end
+
+@Markdown.doc """
+    function SpecOpen(
+      X::Spec{BRT, BRET, RT, RET, MST},
+      I::MPolyLocalizedIdeal{BRT, BRET, RT, RET, MST}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Return the complement of the zero locus of ``I`` in ``X``.
+"""
+function SpecOpen(
+    X::Spec{BRT, BRET, RT, RET, MST},
+    I::MPolyLocalizedIdeal{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
+  base_ring(I) === localized_ring(OO(X)) || error("Ideal does not belong to the correct ring")
+  f = [reduce(f, groebner_basis(localized_modulus(OO(X)))) for f in gens(I)]
+  g = Vector{elem_type(base_ring(OO(X)))}()
+  for a in f
+    iszero(numerator(a)) || (push!(g, numerator(a)))
+  end
+  return SpecOpen(X, g)
+end
+
+function SpecOpen(
+    X::Spec{BRT, BRET, RT, RET, MST},
+    I::MPolyIdeal{RET}
+  ) where {BRT, BRET, RT, RET, MST}
+  return SpecOpen(X, localized_ring(OO(X))(I))
+end
+
+function complement(X::T, Z::T) where {T<:Spec}
+  if !issubset(Z, X) 
+    Z = intersect(X, Z)
+  end
+  return SpecOpen(X, modulus(OO(Z)))
+end
+
+SpecOpen(X::Spec) = SpecOpen(X, [one(base_ring(OO(X)))])
+
+function complement(U::SpecOpen) 
+  if !isdefined(U, :complement)
+    I = radical(saturated_ideal(ideal(localized_ring(OO(ambient(U))), gens(U))))
+    U.complement = subscheme(ambient(U), I)
+  end
+  return U.complement
+end
+
+@Markdown.doc """
+    affine_patches(U::SpecOpen)
+
+Return a list of principal affine open subschemes covering ``U``.
+"""
+function affine_patches(U::SpecOpen)
+  if !isdefined(U, :patches)
+    X = ambient(U)
+    V = Vector{typeof(X)}()
+    for f in gens(U)
+      push!(V, hypersurface_complement(X, f))
+    end
+    U.patches = V
```

Comment:
Perhaps this also works?
```suggestion
    U.patches = [hypersurface_complement(X, f) for f in gens(U)]
```